ls1intum/Artemis
# `Exam mode`: Allow instructors to notify students about programming exam exercise updates and highlight the changes

### Checklist
#### General


#### Client

### Motivation and Context


Instructors currently cannot notify students about changes to the problem statement of an exam programming exercise.
This is possible for all other exercise types and has previously been possible for programming exercises.
Additionally, the changes are not highlighted to be easily detectable for the students and the exercise required a page reload.

### Description
Added the field on the `programming-exercise-update.component` To allow instructors to notify the students.
Added the `exam-exercise-update-highlighter` to the programming-exercise instructions component and modified it to not break problem statements with plantuml diagrams. 
### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 active exam

1. When the exam is running, edit the problem statement and enter some text in the textbox labeled notification about changes
2. As a student you should get a notification at the bottom of the screen and you should see the updated problem statement without reload.
3. As a student you should be able to toggle between highlight differences and show updated version only.
4. Highlighted differences should be legible in both light and dark mode


### Review Progress






#### Performance Review

#### Code Review
#### Manual Tests

### Test Coverage







| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exam-exercise-update-highlighter.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5398/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/exercises/exam-exercise-update-highlighter/exam-exercise-update-highlighter.component.ts.html) | 96.82% | ✅|
| [programming-exercise-instruction.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5398/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/instructions-render/programming-exercise-instruction.component.ts.html) | 88.14% | ✅|
| [exercise-update-notification.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5398/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise-update-notification/exercise-update-notification.component.ts.html) | 100% | ✅|

### Screenshots



https://github.com/ls1intum/Artemis/assets/84102468/5b6a4908-dfd5-4905-93c9-9850b8ac150f


![image](https://github.com/ls1intum/Artemis/assets/84102468/42aa3952-0ea7-4279-a36a-0e390b72cb2d)
![image](https://github.com/ls1intum/Artemis/assets/84102468/ff9b2cbb-176f-472b-9066-5363c32b2888)
![image](https://github.com/ls1intum/Artemis/assets/84102468/88a079fc-e502-43bf-9c2e-88bdd13f2256)
